File: src/content/docs/topics/cs/leetcode-150/graphs/200-number-of-islands.py
def num_islands(grid):
    pass  # TODO: implement

def _run_tests():
    g1 = [['1', '1', '1', '1', '0'], ['1', '1', '0', '1', '0'], ['1', '1', '0', '0', '0'], ['0', '0', '0', '0', '0']]
    assert num_islands(g1) == 1
    g2 = [['1', '1', '0', '0', '0'], ['1', '1', '0', '0', '0'], ['0', '0', '1', '0', '0'], ['0', '0', '0', '1', '1']]
    assert num_islands(g2) == 3
    assert num_islands([]) == 0
    assert num_islands([['1']]) == 1
    assert num_islands([['0']]) == 0
    g3 = [['1', '1'], ['1', '1']]
    assert num_islands(g3) == 1
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    big = [['1'] * 50 for _ in range(50)]
    num_islands(big)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf num-islands 50x50 all-land grid: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/graphs/200-number-of-islands.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}
final class Solution {
    func numIslands(_ grid: [[Character]]) -> Int {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:90546ccc1c6858f851039a36c1b9e7ad005de865846a1f8f43b6605ea27ed65d
    expectEqual(Solution().numIslands([["1", "1", "0", "0", "0"], ["1", "1", "0", "0", "0"], ["0", "0", "1", "0", "0"], ["0", "0", "0", "1", "1"]]), 3, "three-islands")
    expectEqual(Solution().numIslands([["1", "1"], ["1", "1"]]), 1, "connected-land")
    expectEqual(Solution().numIslands([["0"]]), 0, "single-water-cell")
    // EXCLUDED_VECTOR ragged-grid: [[["1","0"],["1"]]] | The problem contract requires a rectangular grid.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
